### PR TITLE
Improve create conditional expressions to handle all expressions

### PR DESCRIPTION
**Pull request by rajyan**
This pull request covers
https://github.com/phpstan/phpstan/issues/8360

*  Solve the infinite loop problem
* Make createConditionalExpressions work for all expressions

improves result of https://github.com/phpstan/phpstan/issues/4173 but not completely fixed yet

### CODE REVIEW

**Comment by rajyan** [comment]
Issue bot seems not working well. Looks like a cache problem?
This change should notify issue-7705 and issue-4173 changes like in
https://github.com/phpstan/phpstan-src/pull/1950#issuecomment-1307318676

**Comment by staabm** [comment]
> Issue bot seems not working well. Looks like a cache problem?
> This change should notify issue-7705 and issue-4173 changes like in

hmm I have the impression it works:

https://github.com/phpstan/phpstan-src/actions/runs/3488347704

grafik

**Comment by rajyan** [comment]
Oh, thank you for noticing! Not sure what I was looking at...

**Comment by rajyan** [comment]
@ondrejmirtes 
I think the result of issue bot is wrong when there is a infinite loop. issue-7705 was causing an infinite loop before https://github.com/phpstan/phpstan-src/pull/2007/commits/dd478c0b5b1999856fab94c1486235060eb4261f , but the issue bot was saying no error.

**Comment by herndlm** [comment]
Loops are generalised and skipped after a couple of iterations AFAIK. Could that be related?

**Comment by rajyan** [comment]
Not that "loop" I think.
This change causes something like
Resolve conditionalExpressionType of `$foo` => Need to `getType` of contional `$isFoo || $isBar` => Resolve type of `$isFoo || $isBar`  => `getType` of BooleanOr uses `filterByFalsyScope` => In `filterByFalsyScope` conditionalExpressionType of `$foo` is evaluate again...

**Comment by herndlm** [comment]
Ah sorry, yeah, completely wrong loop I was thinking about.. xD

**Comment by ondrejmirtes** [comment]
Thank you!

File: tests/PHPStan/Rules/Variables/data/bug-4173.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+			throw new \InvalidArgumentException();
+		}
+
+		$field = trim($field);
+
+		if (count($row) === 1) {
+			return $field;
+		}
+
+		if (count($row) === 2) {
```
Comment (mvorisek) [inline tests/PHPStan/Rules/Variables/data/bug-4173.php:25]: This PR seems to cache result of function result 1:1 and retrive the result from cache later. I observed this in phpstan v 1.9.2 (the latest rel).

Is that desired? For `count()` it is legit and thank you. However for expression like `TraitUtil::hasTrait($object, DebugTrait::class)` (a custom function) how does php know/guarantee the result will be the same, does it reason it based on the `TraitUtil::hasTrait` impl or just guess it when the passed args are the same?

Comment (rajyan) [reply tests/PHPStan/Rules/Variables/data/bug-4173.php:25]: Yes this is intended

> how does php know/guarantee the result will be the same

When
* TraitUtil::hasTrait is not marked impure
* The exact same `TraitUtil::hasTrait($object, DebugTrait::class)` is called
* $object is not invalidated

> just guess it when the passed args are the same?

so it's something like that.

Comment (rajyan) [reply tests/PHPStan/Rules/Variables/data/bug-4173.php:25]: By the way, remembering function results was working before this PR already.
https://phpstan.org/blog/remembering-and-forgetting-returned-values
this PR made them work for more complex conditions too.

Comment (mvorisek) [reply tests/PHPStan/Rules/Variables/data/bug-4173.php:25]: Thank you, I did not know user functions are "pure"/remembered by default.

Comment (staabm) [reply tests/PHPStan/Rules/Variables/data/bug-4173.php:25]: See https://phpstan.org/blog/remembering-and-forgetting-returned-values

**Pull request merged**